perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3030,22 +3045,104 @@ private void categorizeMembersForLightweightSynchronization(PerunSession sess, G
+					membersToRemove.remove(richMember);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Checks if userExtSource of candidate corresponds to some userExtSource of RichMember already in group. Whether
+	 * it's by extLogin or value of ues attribute. Returns this RichMember if it does, null otherwise.
+	 *
+	 * @param sess perun session
+	 * @param key ues of candidate
+	 * @param mappingStructure map of ues to richMembers in group
+	 * @return richMember or null
+	 */
+	private RichMember processRichUserExtSourceDuringCategorization(PerunSession sess, RichUserExtSource key, Map<UserExtSource, RichMember> mappingStructure) {
+		//check if ues is in the map
+		if (mappingStructure.containsKey(key.asUserExtSource())) {
+			return mappingStructure.get(key.asUserExtSource());
+		}
+
+		for (Attribute attribute : key.getAttributes()) {
+
+			//logins are either single value string, or multi value list
+			List<String> loginsFromAttribute;
+			if (attribute.getType().equals(String.class.getName())) {
+				loginsFromAttribute = Arrays.asList(attribute.valueAsString());
+			} else {
+				loginsFromAttribute = attribute.valueAsList();
+			}
+
+			//check if extSource with login from ues attribute is in the map
+			for (UserExtSource ues : mappingStructure.keySet()) {
+				if (ues.getExtSource().equals(key.asUserExtSource().getExtSource())) {
+					try {
+						Attribute attributeToCompare = perunBl.getAttributesManagerBl().getAttribute(sess, ues, attribute.getName());
+
+						List<String> attributeToCompareValue;
+						if (attributeToCompare.getType().equals(String.class.getName())) {
+							attributeToCompareValue = Arrays.asList(attribute.valueAsString());
+						} else {
+							attributeToCompareValue = attribute.valueAsList();
+						}
+
+						//calculate intersection of attribute values
+						List<String> result = attributeToCompareValue.stream().distinct().filter(loginsFromAttribute::contains).collect(Collectors.toList());
+
+						if (!result.isEmpty()) {
+							return mappingStructure.get(ues);
+						}
+					} catch (WrongAttributeAssignmentException | AttributeNotExistsException e) {
+						//attribute not found, just continue
+					}
+				}
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Updates (or creates) ues attributes to additional userExtSources of candidate.
+	 *
+	 * @param sess perun session
+	 * @param candidateSync candidate whose ues attributes are being updated
+	 */
+	private void updateAdditionalUserExtSourceAttributes(PerunSession sess, CandidateSync candidateSync) {
+		for (RichUserExtSource ues : candidateSync.getRichUserExtSources()) {
+			for (Attribute attribute : ues.getAttributes()) {
+				try {
+					Attribute attributeToAdd = perunBl.getAttributesManagerBl().getAttribute(sess, ues.asUserExtSource(), attribute.getName());
+
+					if (attributeToAdd.getValue() == null || !attributeToAdd.getValue().equals(attribute.getValue())) {
+						perunBl.getAttributesManagerBl().setAttribute(sess, ues.asUserExtSource(), attribute);
+					}
```

Comment:
I am afraid this is not ok. We need to respect what is already set in Perun. This means if there is one value attribute (String type), and it is already set with a different value, you need to skip this user and inform about it to the messages (variable 'skippedMembers'). If the value of the attribute is an array list, you can update the attribute by missing values.